# ibdtools

## Compilation

1. Dependencies: The code relies on zlib, htslib and tbb libraries and  a version gcc
   compilers that supports C++ 17 standard. They can be installed by conda.

```sh
conda install -c conda-forge gxx_linux-64 htslib tbb-devel 
```

2. Adjust the paths for libraries: `hts`, `zlib` and `tbb` in the `Makefile`; 

3. Add these lib paths to the `LD_LIBRARY_PATH` environment variable if not found in
   default paths.

4. Run `make` at the project root directory to compile. The binaries will be under
   `build/`.


## `ibdsort`

```
Usage: ibdsort [OPTION...] IBD_FILE
ibdsort prepares formatted and sorted IBD data for ibdmerge. It uses external merge sort
algorithm and is written to handle large datasets. 

 Required arguments:
  IBD_FILE                   Input IBD file from hap_ibd for a single
                             chromosome

 Optional arguments:
  -k, --k_ways=integer       Number of ways for merging. Default 8
  -K, --keep_tmp_file        Keep all tmp file for debuging
  -m, --memory=float         Estimated maximum memory in Gb to use. Default 1
  -o, --out=filename         IBD output file after formatting and sorting,
                             default stdout
  -p, --tmp_file_prefix=string   The tmp file prefix. The file name will be
                             {tmp_file_prefix}_{xx}. Default is using the
                             IBD_FILE.
  -s, --sample=filename      Sample file. If provided, IBD records will be
                             sorted according to the alphabetical order of
                             sample name; if not provodied, IBD samples will be
                             sorted according to the order the program first
                             encounters the sample name in the IBD records
                             Default: NULL
  -v, --verbose              Show processing updates, default false

  -?, --help                 Give this help list
      --usage                Give a short usage message
  -V, --version              Print program version

```

## `ibdmerge`

```
Usage: ibdmerge [OPTION...] IBD_FILE VCF_FILE MAP_FILE
ibdmerge merges IBD segments following Browning's tool merge-ibd-segments.jar.
This is written to handle large datasets.

 Required arguments:
  IBD_FILE                   IBD file with a format of  >=4 columns delimited
                             by tabs.The first 4 columns are sn1, sn2,
                             start_bp, end_bp. The file can be generated by
                             ibdsort
  MAP_FILE                   MAP file (space delimited) used to call IBD
  VCF_FILE                   VCF file (tab delimited) used to call IBD. The VCF
                             file is expected to only have biallelic sites

 Optional arguments:
  -d, --discord=int          max number of genotypes in IBD gap that are
                             inconsistent with IBD, default 1
  -m, --max_cM=float         max length of gap between IBD segments (cM),
                             default 0.6
  -o, --out=filename         IBD output file after merging, default stdout
  -v, --verbose              Show processing updates, default false

  -?, --help                 Give this help list
      --usage                Give a short usage message
  -V, --version              Print program version
```

## `ibdtotal`

```
Usage: ibdtotal [OPTION...] SAMPLE_FILE OUT_FILE
Summarize sample-pair total ibd and save it to a binary matrix.
Total IBD length is converted to integers with a unit of 0.1cM.

 Required arguments:
  OUT_FILE                   binary file to save the lower trigular total IBD
                             matrix
  SAMPLE_FILE                sample file that list all samples in IBD files,
                             one line per sample

 Optional arguments: at least one of -i or -m should be specified
  -i, --IBD_file=FILE        IBD file with a format of  5 columns delimited by
                             tabs. Columns are sn1, sn2, start_bp, end_bp,
                             length_in_cM
  -m, --Matrix_files=FILE [FILE...]
                             Existing Ibd matrix file to combine; can be used
                             to merge with partial total IBD matrix calculated
                             from other chromosomes. ATTENTION: -m option
                             should be specified at the END!!
```
